crates/wasi-common/src/table.rs

```diff
@@ -0,0 +1,111 @@
+use crate::{Error, ErrorExt};
+use std::any::Any;
+use std::cell::{Ref, RefCell, RefMut};
+use std::collections::HashMap;
+
+/// The `Table` type is designed to map u32 handles to resources. The table is now part of the
+/// public interface to a `WasiCtx` - it is reference counted so that it can be shared beyond a
+/// `WasiCtx` with other WASI proposals (e.g. `wasi-crypto` and `wasi-nn`) to manage their
+/// resources. Elements in the `Table` are `Any` typed.
+///
+/// The `Table` type is intended to model how the Interface Types concept of Resources is shaping
+/// up. Right now it is just an approximation.
+pub struct Table {
+    map: HashMap<u32, RefCell<Box<dyn Any>>>,
```

Comment:
What are the thread-safety guaranty requirements?

Couldn't multiple threads access the same key concurrently?

Instead of `Any`, wouldn't it be safer to enforce `Any + Send + Sync`?

Reply:
A `Table` should only be associated with a single `wasmtime::Store`, which is not Send or Sync. So, I made no attempt to provide any thread-safety guarantees.

Reply:
So a handle will never be shared by multiple threads?

Is this a requirement of the WebAssembly threading proposal, or a temporary limitation in Wasmtime?

More generally, should hostcalls assume that a resource will never be shared?

Reply:
Right, a handle will never be shared between threads. It is a limitation of Wasmtime until we choose to implement a threading proposal.

If a hostcall needs to assume a resource is shared between multiple Stores running in separate threads, it can do its own internal arc/mutex wrapping inside the resource.